src/Core/Config/Util/ConfigFileManager.php

```diff
@@ -177,49 +182,83 @@ private function loadDataConfig(Cache $config)
 			$content = '<?php return [];';
 
 			$configStream = fopen($filename, 'r');
-			if (flock($configStream, LOCK_SH)) {
-				$content = fread($configStream, filesize($filename));
-				if (!$content) {
-					throw new ConfigFileException(sprintf('Couldn\'t read file %s', $filename));
+			try {
+				if (flock($configStream, LOCK_SH)) {
+					$content = fread($configStream, filesize($filename));
+					if (!$content) {
+						throw new ConfigFileException(sprintf('Couldn\'t read file %s', $filename));
+					}
 				}
+			} finally {
 				flock($configStream, LOCK_UN);
+				fclose($configStream);
 			}
 
-			fclose($configStream);
-
 			$dataArray = eval('?>' . $content);
 
-			if (!is_array($dataArray)) {
-				throw new ConfigFileException(sprintf('Error loading config file %s', $filename));
+			if (is_array($dataArray)) {
+				$configCache->load($dataArray, Cache::SOURCE_DATA);
 			}
-
-			$config->load($dataArray, Cache::SOURCE_DATA);
 		}
 	}
 
 	/**
 	 * Saves overridden config entries back into the data.config.phpR
 	 *
-	 * @param Cache $config The config cache
+	 * @param Cache $configCache The config cache
 	 *
 	 * @throws ConfigFileException In case the config file isn't writeable or the data is invalid
 	 */
-	public function saveData(Cache $config)
+	public function saveData(Cache $configCache)
 	{
-		$data = $config->getDataBySource(Cache::SOURCE_DATA);
-
-		$encodedData = ConfigFileTransformer::encode($data);
+		$filename = $this->configDir . '/' . self::CONFIG_DATA_FILE;
 
-		if (!$encodedData) {
-			throw new ConfigFileException('config source cannot get encoded');
+		if (file_exists($filename)) {
+			$fileExists = true;
+		} else {
+			$fileExists = false;
 		}
 
-		$configStream = fopen($this->configDir . '/' . self::CONFIG_DATA_FILE, 'w');
+		$configStream = fopen($filename, 'c+');
```

Comment:
Please vary the flag depending on `$fileExists` to replace some of the following convolution.

Reply:
I read about a possible, very very small, race condition during this check, so two `save()` will try both a `w+` instead a `w+` and `r+` lock. That's why I let it the OS decide what to do :-)

Reply:
Ok, I feel like this block deserves a lengthy explanation in a multi-line comment.

Reply:
Aaaand here's the doc :-)

But plz have a look at my english grammer/syntax, because it's hard for me as non-native speaker (and tbh I'm frequently writing english only for Friendica *gg*)